### PR TITLE
Remove field traits from behind `std` feature flag

### LINKED ISSUE #25: Add `alloc` feature flag
In #4 we introduced a `std` feature flag, to enable downstream users to disable anything relying on `std`. This happened to gate all of the traits, because both `FieldExt` and `CurveAffine` depended on `std::io`.

We removed `std::io` usage from `FieldExt` in #20, so it could now be removed from any feature flag. `CurveAffine` still relies on `std::io`, but I want to remove that as well. Then we're just left with the fact that `CurveExt` requires `FieldExt`, `FieldExt` is bound on `SqrtRatio`, and the only way we can implement `SqrtRatio` right now depends on `SqrtTables`, which requires `alloc`.

So, we should add an `alloc` feature flag, and move everything behind it. We might also be able to remove the `std` feature flag as a result (which has not been in a release yet). There's no point in having the traits themselves be no-std accessible if we can't implement them without `alloc`. Once we have a fallback for `SqrtRatio` we can consider making the other traits no-std usable.

### PR DESCRIPTION
**Pull request by str4d**
Closes #25.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -21,6 +21,7 @@ and this project adheres to Rust's notion of
   - `FieldExt::from_u64` (use `From<u64> for ff::PrimeField` instead).
   - `FieldExt::{from_bytes, read, to_bytes, write}`
     (use `ff::PrimeField::{from_repr, to_repr}` instead).
+  - `FieldExt::rand` (use `ff::Field::random` instead).
```
Comment (daira) [inline CHANGELOG.md:25]: Also document the changes to `std` gating.

**Review by daira — APPROVED** [review]
utACK

File: src/arithmetic/fields.rs [hunk 2]
```diff
@@ -54,7 +53,40 @@ pub trait SqrtRatio: ff::PrimeField {
...
+        // a = num * inv0(div)
+        //   = {    0    if div is zero
+        //     { num/div otherwise
+        //
+        // b = G_S * a
+        //   = {      0      if div is zero
+        //     { G_S*num/div otherwise
+        //
+        // Since G_S is non-square, a and b are either both zero (and both square), or
+        // only one them is square. We can therefore choose the square root to return
```
Comment (daira) [inline src/arithmetic/fields.rs:68]:
```suggestion
        // only one of them is square. We can therefore choose the square root to return
```

File: CHANGELOG.md (hunk 1 above)
Comment (str4d) [reply CHANGELOG.md:25]: We've not released a crate version with the `std` feature flag, so to crate users they'll see the addition of no-std support. Agreed on documenting the new flags, but I'll do that in a separate PR once everything is stable.

**Pull request merged**